src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,197 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+            StorageItem item = snapshot.GetAndChange(key);
+            int value = (int)(BigInteger)item;
+            item.Add(1);
+            return value;
+        }
+
+        public static async Task<(ContractState contractState, IReadOnlyList<(string key, string value)> storagePairs)> DownloadContractStateAsync(
+                string contractHash, string rpcUri, uint stateHeight)
+        {
+            if (!UInt160.TryParse(contractHash, out var _contractHash))
+            {
+                throw new ArgumentException($"Invalid contract hash: \"{contractHash}\"");
+            }
+
+            if (!TransactionExecutor.TryParseRpcUri(rpcUri, out var uri))
+            {
+                throw new ArgumentException($"Invalid RpcUri value \"{rpcUri}\"");
+            }
+
+            using var rpcClient = new RpcClient(uri);
+            var stateAPI = new StateAPI(rpcClient);
+
+            uint height = stateHeight;
+            if (height == 0)
+            {
+                uint? localRootIndex;
+                try
+                {
+                    (localRootIndex, _) = await stateAPI.GetStateHeightAsync();
+                }
+                catch (RpcException e)
+                {
+                    if (e.Message.Contains("Method not found"))
+                    {
+                        throw new Exception(
+                            "Could not get state information. Make sure the remote RPC server has state service support");
+                    }
+                    throw;
+                }
+
+                height = localRootIndex.HasValue ? localRootIndex.Value
+                    : throw new Exception($"Null \"{nameof(localRootIndex)}\" in state height response");
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+
+            // rpcClient.GetContractStateAsync returns the current ContractState, but this method needs
+            // the ContractState as it was at stateHeight. ContractManagement stores ContractState by
+            // contractHash with the prefix 8. The following code uses stateAPI.GetStateAsync to retrieve
+            // the value with that key at the height state root and then deserializes it into a ContractState
+            // instance via GetInteroperable.
+
+            var key = new byte[21];
+            key[0] = 8; // ContractManagement.Prefix_Contract
+            _contractHash.ToArray().CopyTo(key, 1);
+
+            var contractStateBuffer = await stateAPI.GetStateAsync(
+                stateRoot.RootHash, NativeContract.ContractManagement.Hash, key);
+            var contractState = new StorageItem(contractStateBuffer).GetInteroperable<ContractState>();
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, default);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(NeoSystem neoSystem, ContractState state,
```

Comment:
How should we handle the user trying to download a native contract? Obviously, we can't support downloading native contract state, but we could support downloading native contract storage. Not sure it's a good idea tho.